gnupg-pkcs11-scd/command.c

```diff
@@ -1324,18 +1324,39 @@ gpg_error_t cmd_pkdecrypt (assuan_context_t ctx, char *line)
 	 */
 	_data.data = data->data;
 	_data.size = data->size;
-	if (
+	
+	if (_data.size >= (128-16) && _data.size < 128) /* 1024 bit key */
+		fixuplen = 128 - _data.size;
+	else if (_data.size >= (192-16) && _data.size < 192) /* 1536 bit key.  */
+        	fixuplen = 192 - _data.size;
+    	else if (_data.size >= (256-16) && _data.size < 256) /* 2048 bit key.  */
+        	fixuplen = 256 - _data.size;
+    	else if (_data.size >= (384-16) && _data.size < 384) /* 3072 bit key.  */
+        	fixuplen = 384 - _data.size;
+    	else if (_data.size >= (512-16) && _data.size < 512) /* 4096 bit key.  */
+        	fixuplen = 512 - _data.size;
+	else if (
 		*_data.data == 0 && (
 			_data.size == 129 ||
 			_data.size == 193 ||
 			_data.size == 257 ||
 			_data.size == 385 ||
-			_data.size == 513
-		)
-	) {
+			_data.size == 513)) 
+	{
 		_data.data++;
 		_data.size--;
```

Comment:
we repeat these numbers over and over, we need to optimize...

something like:

```c
static const int keysizes[] = {1024/8, 1536/8, ...., 0};
char *_dataptr = NULL;
int *s;
for (s = keysizes; *s != 0; s++) {
    if (_data.size >= *s - 16 && _data.size < *s) {
        int fixuplen = *s - _data.size;
        dataptr = calloc(fixuplen + _data.size);
        memcpy(dataptr + fixuplen, _data.data, _data.size);
        _data.data = dataptr;
        _data.size += fixuplen;
        break;
    }
    if (*_data.data == 0 && _data.size == *s+1) {
        _data.data++;
        _data.size--;
       break;
    }
}

...

cleanup:
if (dataptr != NULL) {
    free(dataptr);
    dataptr = NULL;
}
```